source/src/encryption/cs_KeysAndAccess.cpp

```diff
@@ -100,21 +102,32 @@ bool KeysAndAccess::getKey(EncryptionAccessLevel accessLevel, buffer_ptr_t outBu
 }
 
 cs_data_t KeysAndAccess::getSetupKey() {
-	KeysAndAccessDebug("getSetupKey valid=%u", _setupKeyValid);
+	LOGKeysAndAccessDebug("getSetupKey valid=%u", _setupKeyValid);
 	if (!_setupKeyValid) {
 		return cs_data_t();
 	}
 	return cs_data_t(_setupKey, sizeof(_setupKey));
 }
 
 void KeysAndAccess::generateSetupKey() {
-	KeysAndAccessDebug("generateSetupKey opMode=%u", _operationMode);
+	LOGKeysAndAccessDebug("generateSetupKey opMode=%u", _operationMode);
 	if (_operationMode == OperationMode::OPERATION_MODE_SETUP) {
 		RNG::fillBuffer(_setupKey, SOC_ECB_KEY_LENGTH);
 		_setupKeyValid = true;
 	}
 }
 
+cs_ret_code_t KeysAndAccess::setSetupKey(cs_data_t data) {
+	LOGKeysAndAccessDebug("Set setup key");
+	if (data.len != sizeof(_setupKey)) {
+		LOGw("Wrong key length: %u", data.len);
+		return ERR_WRONG_PAYLOAD_LENGTH;
+	}
```

Comment:
Having this method might not be really cool. It's called in CrownstoneCentral on incoming connection. Does that class have the access rights?

Reply:
SetupKey is a temp key, it changes each connection. So it's fine if the CrownstoneCentral sets it for an outgoing connection.